packages/yarnpkg-pnpify/sources/Hoister.ts

```diff
@@ -0,0 +1,289 @@
+
+import fs                                                             from 'fs';
+
+import {RawHoister, ReadonlyPackageTree, PackageId, PackageInfo}      from './RawHoister';
+import {HoistedTree, PackageTree}                                     from './RawHoister';
+
+// Babel doesn't support const enums, thats why we use non-const enum for LinkType in @yarnpkg/pnp
+// But because of this TypeScript requires @yarnpkg/pnp during runtime
+// To prevent this we redeclare LinkType enum here, to not depend on @yarnpkg/pnp during runtime
+export enum LinkType {HARD = 'HARD', SOFT = 'SOFT'};
+
+type PnpWalkApi = Pick<PnpApi, 'getPackageInformation' | 'getDependencyTreeRoots' | 'topLevel' | 'resolveVirtual'>;
+
+/**
+ * Node modules tree.
+ */
+export type NodeModulesTree = Map<PortablePath, Set<Filename> | [PortablePath, LinkType]>;
+
+export interface HoisterOptions {
+  optimizeSizeOnDisk?: boolean;
+  pnpifyFs?: boolean;
+}
+
+/** node_modules path segment */
+const NODE_MODULES = toFilename(`node_modules`);
+const NODE_MODULES_SUFFIX = `/node_modules`;
+
+/** Package locator key for usage inside maps */
+type LocatorKey = string;
+
+/**
+ * Retrieves full package list from PnP API and builds hoisted `node_modules` directories
+ * representation in-memory.
+ */
+export class Hoister {
+  /** Raw hoister which does all the hoisting heavy lifting */
+  private readonly rawHoister = new RawHoister();
+
+  /** Hoister options */
+  private readonly options: HoisterOptions;
+
+  public constructor(options: HoisterOptions = {optimizeSizeOnDisk: false}) {
+    this.options = options;
+  }
+
+  /**
+   * Traverses PnP tree and produces input for the `RawHoister`
+   *
+   * @param pnp PnP API
+   *
+   * @returns package tree, packages info and locators
+   */
+  private buildPackageTree(pnp: PnpWalkApi): { packageTree: ReadonlyPackageTree, packages: PackageInfo[], locators: PackageLocator[] } {
+    const packageTree: PackageTree = [];
+    const packages: PackageInfo[] = [];
+    const locators: PackageLocator[] = [];
+    const locatorToPackageMap = new Map<LocatorKey, PackageId>();
+    const packageInfos: PackageInformation<NativePath>[] = [];
+
+    const pnpRoots = pnp.getDependencyTreeRoots ? pnp.getDependencyTreeRoots() : [pnp.topLevel];
+
+    let lastPkgId = 0;
+
+    const getLocatorKey = (locator: PackageLocator): LocatorKey => `${locator.name}:${locator.reference}`;
+
+    const assignPackageId = (locator: PackageLocator, pkg: PackageInformation<NativePath>) => {
+      const locatorKey = getLocatorKey(locator);
+      let pkgId = locatorToPackageMap.get(locatorKey);
+      if (!pkgId) {
+        pkgId = lastPkgId;
+        lastPkgId++;
+        const packagePath = npath.toPortablePath(pkg.packageLocation);
+        let weight = 1;
+        if (this.options.optimizeSizeOnDisk && packagePath.indexOf(`.zip${NODE_MODULES_SUFFIX}`) > 0)
+          weight = fs.statSync(packagePath.split(NODE_MODULES_SUFFIX)[0]).size;
```

Comment:
Could this be abstracted in a `getPackageWeight` helper function?

```ts
function getPackageWeight(packagePath, {optimizeSizeOnDisk}) {
  if (!optimizeSizeOnDisk)
    return 1;

  const archivePath = getArchivePath(packagePath);
  if (archivePath === null)
    return 1;

  return xfs.statSync(archivePath).size;
}
```

Reply:
Done